### ISSUE TITLE
Programmatic service registration

### ISSUE THREAD

**Issue by lburgazzoli** [spring-cloud/spring-cloud-commons#361]
According to the  [documentation](https://github.com/spring-cloud/spring-cloud-commons/blob/master/docs/src/main/asciidoc/spring-cloud-commons.adoc#serviceregistry), `ServiceRegistry` provides methods to register custom services but it does not appear to work as documented:

I have a custom [Registration](ttps://github.com/lburgazzoli/atomix-boot/blob/master/atomix-boot-starter-node/src/main/java/com/github/lburgazzoli/atomix/boot/node/AtomixBootNodeRegistration.java) but when I invoke [ServiceRegistry::register](https://github.com/lburgazzoli/atomix-boot/blob/master/atomix-boot-starter-node/src/main/java/com/github/lburgazzoli/atomix/boot/node/AtomixBootNode.java#L67), an exception is thrown:

```
java.lang.ClassCastException: com.github.lburgazzoli.atomix.boot.node.AtomixBootNodeRegistration cannot be cast to org.springframework.cloud.consul.serviceregistry.ConsulRegistration
at org.springframework.cloud.consul.serviceregistry.ConsulServiceRegistry.register(ConsulServiceRegistry.java:41) ~[spring-cloud-consul-discovery-2.0.0.RC1.jar:2.0.0.RC1]
at com.github.lburgazzoli.atomix.boot.node.AtomixBootNode.lambda$start$0(AtomixBootNode.java:67) ~[classes/:na]
...
```

Dependencies:
- sring-boot 2.0.1
- spring-cloud 2.0.0.RC1
- spring-cloud-consul 2.0.0.RC1

**Comment by ryanjbaxter** [comment]
Sounds like this might belong in https://github.com/spring-cloud/spring-cloud-consul

**Comment by lburgazzoli** [comment]
@ryanjbaxter 

I do not think so, also spring-cloud-zookeeper and spring-cloud-netflix have an implementation that contradicts the documentation, see:

- https://github.com/spring-cloud/spring-cloud-zookeeper/blob/master/spring-cloud-zookeeper-discovery/src/main/java/org/springframework/cloud/zookeeper/serviceregistry/ZookeeperServiceRegistry.java#L65
- https://github.com/spring-cloud/spring-cloud-netflix/blob/master/spring-cloud-netflix-eureka-client/src/main/java/org/springframework/cloud/netflix/eureka/serviceregistry/EurekaServiceRegistry.java#L38

**Comment by lburgazzoli** [comment]
So this means that I do not have a way to register a custom service in a "cross-platform" way, correct ? Seems a limitation to me

**Comment by spencergibb** [comment]
They each require system specific objects. Not sure how to get around that.

**Comment by lburgazzoli** [comment]
I understand but it is not possible to create a ConsulRegistrationas it requires an instance of NewService which of course I do not have and it should not be required otherwhise it is not possible to implement a portable service.

**Comment by spencergibb** [comment]
It was never intended to be able to reuse a `Registration` across different service registries.

**Comment by spencergibb** [comment]
How would it know what to register without a `NewService`?

**Comment by lburgazzoli** [comment]
But the documentation does not say that, it says the opposite

**Comment by spencergibb** [comment]
And we will change that.

**Comment by lburgazzoli** [comment]
In consul I can registry any kind of service, I have to expose the host an port to which the service answer which does not depend on consul.

**Comment by spencergibb** [comment]
Spring cloud consul interacts with consul using `consul-api` which is where `NewService` comes from. There's no way around that.

**Comment by lburgazzoli** [comment]
Well, you can wrap any plain Registration in a ConsulServiceRegistration isn't it ?

**Comment by spencergibb** [comment]
And how would the `ConsulRegistration` get the information it needs from the wrapped `Registration`?

**Comment by lburgazzoli** [comment]
What consul/zookeeper/eureka specific info it needs ?

**Comment by spencergibb** [comment]
specific classes populated correctly for each one.

**Comment by lburgazzoli** [comment]
NewService can be constructed out of a Registration imho

**Comment by lburgazzoli** [comment]
Said that I can work around it by writing specific implementation and leverage conditional beans but feels  limited to me as I need to write a custom adapter for each spring-cloud-xyz

**Comment by ryanjbaxter** [comment]
Closed via 88d3439
